endpoint.go

```diff
@@ -122,17 +121,39 @@ type InEndpoint struct {
 	*endpoint
 }
 
-// Read reads data from an IN endpoint.
+// Read reads data from an IN endpoint. Read returns number of bytes obtained
+// from the endpoint. Read may return non-zero length even if
+// the returned error is not nil (partial read).
 func (e *InEndpoint) Read(buf []byte) (int, error) {
-	return e.transfer(buf)
+	return e.transfer(context.Background(), buf)
+}
+
+// ReadContext reads data from an IN endpoint. Read returns number of bytes
+// obtained from the endpoint. Read may return non-zero length even if the
+// returned error is not nil (partial read).
+// The passed context can be used to control the cancellation of the read. If
+// the context is cancelled, ReadContext will return a TransferTimedOut error.
```

Comment:
Contexts can be cancelled due to timeout, but also explicitly by the user. It seems slightly cleaner to me to distinguish the two cases, instead of unconditionally returning a TransferTimedOut.

Reply:
done

Reply:
You can distinguish between the two by using https://golang.org/pkg/context/#pkg-variables

If that’s not possible, I would prefer TransferCancelled over TransferTimedOut.

Reply:
I could, although now I think it creates additional confusion, since the error returned is not the transfer error. I've updated the code to just call cancel() on cancelled context, and the return error now automatically propagates from the transfer return status.
Also this fixes the partial read/write that might have happened with cancelled context, now explicitly tested.